### ISSUE TITLE
🚀 [Feature] Compress uploaded images automatically

### ISSUE THREAD

**Issue by alina-beck** [Human-Connection/Human-Connection#1922]
## :rocket: Feature
Post authors should be able to upload images without too much hassle (so the file size should not be too heavily restricted).

Users should be able to view posts without having to download huge images (eating up bandwidth and taking a long time).

So the desired solution is to automatically compress all images that are uploaded to our network by our users. From my experience I would propose a maximum size of roughly 500-800KB – though this should be tested to make sure the images still have an acceptable quality.

### Additional context
This came up when we realized that images touched by `cropperjs` would have file sizes around 10MB! More thoughts here: #1933

**Comment by mattwr18** [comment]
I have been playing around with this in https://github.com/Human-Connection/Human-Connection/pull/1935 and learned that the `canvas.toBlob()` that we use automatically uses `loosy compression`, which I had to look up what means.

Point, being, after I specified type `image/jpeg` as @vardyb suggested, and created a new File from the blob to restore filename, the cropped images saved are anywhere between around `175k-950k`, kinda in line with what you suggest, no @alina-beck 

Could you have a look at the `PR`?

**Comment by alina-beck** [comment]
@mattwr18 I approved your PR! Great that you figured this out! :smile:

I'm wondering if we should still leave this issue open... File size obviously isn't a problem anymore as long as we force users to open the cropping tool, which we currently do but were thinking of changing to an opt-in feature soon. When that is implemented cropped images will have the desired file size but others not. So, what do you think?

**Comment by mattwr18** [comment]
I think that makes sense

**Comment by alina-beck** [comment]
@HelixX23 had an interesting comment about image validation and hidden content in images – we could consider this when working on the above issue: https://github.com/Human-Connection/Human-Connection/issues/2119#issuecomment-554490773

**Comment by alina-beck** [comment]
still relevant – and probably related to #2255

**Comment by HelixX23** [comment]
Maybe imagemagick could do the trick? 
https://www.imagemagick.org/script/index.php

**Comment by mattwr18** [comment]
So, we have added this to the `EPIC` #2265 and therefore it's been put into the current Sprint, but we haven't had much time to plan it out, and share ideas... 
I opened a `PR` for an Image microservice, which leverages `sharp` 
https://github.com/lovell/sharp
https://github.com/Human-Connection/Human-Connection/pull/2602/files , but this seems to be more about resizing images, then compressing them.

I have seen this library, imagemin https://github.com/imagemin/imagemin, which has lots of stars and seems to be maintained. It takes an array of files/ or glob patterns and uses a list of plugins to compress the images.

Would we add this to a middleware? should we write them, so that the future requests could return the compressed images?? Insights? Direction?

**Comment by mattwr18** [comment]
I have ran the command locally with the `imagemin-cli` tool and here are the results
before:
```
du -h public/uploads/1574331493268-IMG_20191002_170522
8,4M	public/uploads/1574331493268-IMG_20191002_170522
```
after this command `yarn compress:image public/uploads/1574331493268-IMG_20191002_170522 > public/uploads/yummy-sandwich-optimized`
```
du -h public/uploads/yummy-sandwich-optimized 
4,5M	public/uploads/yummy-sandwich-optimized
```
it's a bit bigger than half the size...

**Comment by roschaefer** [comment]
Well, certainly we don't want to convert everything to .jpg. Think of logos with sharp edges which are best saved as .png. JPG is a not a lossless compression and will lead to [odd artefacts around edges](https://www.google.com/search?q=jpeg+compression+artefacts&tbm=isch).

I'm leaning against automatic compression of images *if we can provide smaller versions* of them. Those smaller sizes should essentially solve our problems. The only problem that we would solve here with automatic compression in the backend is disk size, or am I missing something?

**Comment by mattwr18** [comment]
Yeah, I am sure that we don't want to convert all to `jpg`, but that is what we are currently doing. A benefit of using this package is that we can compress without needing to convert.

I'll wait till after the meeting to see which direction we go in, and we can discuss whether to leave this open till we are using `spaces`, close it, or combine them.

**Comment by alina-beck** [comment]
I'm in favor of not converting images to `jpg` but I do think we should compress images – even if we start providing smaller file sizes to the frontend. Or instead set a file size limit for uploads.

It sure won't be necessary for most uploads and I know disc space is not _that_ expensive but think of someone exploiting this and uploading images that have file sizes of several `GB`.

**Comment by Mogge** [comment]
not stale
